### PR TITLE
GH-39156: [C++][Compute] Fix negative duration division

### LINKED ISSUE #39156: [C++] Division with negative duration types leads to `NaN`.
### Describe the bug, including details regarding any error messages, version, and platform.

```python
import pyarrow as pa

pos_one_second = pa.scalar(1, type=pa.duration("us"))  # 1s
pos_two_second = pa.scalar(2, type=pa.duration("us"))  # 2s
neg_one_second = pa.scalar(-1, type=pa.duration("us"))  # -1s
neg_two_second = pa.scalar(-2, type=pa.duration("us"))  # -2s

print(pa.compute.divide(pos_one_second, pos_one_second))  # 1.0 ✔
print(pa.compute.divide(pos_two_second, pos_one_second))  # 2.0 ✔
print(pa.compute.divide(pos_one_second, pos_two_second))  # 0.5 ✔
print(pa.compute.divide(pos_two_second, pos_two_second))  # 1.0 ✔

print(pa.compute.divide(pos_one_second, pos_one_second))  # 1.0 ✔
print(pa.compute.divide(neg_one_second, pos_one_second))  # NAN ✘
print(pa.compute.divide(pos_one_second, neg_one_second))  # NAN ✘
print(pa.compute.divide(neg_one_second, neg_one_second))  # NAN ✘

print(pa.compute.divide(pos_one_second, pos_one_second))  # 1.0 ✔
print(pa.compute.divide(neg_two_second, pos_one_second))  # NAN ✘
print(pa.compute.divide(pos_one_second, neg_two_second))  # NAN ✘
print(pa.compute.divide(neg_two_second, neg_two_second))  # NAN ✘
```

Tested with 14.0.1, first noted here: https://github.com/pandas-dev/pandas/issues/56409

Related:

https://github.com/apache/arrow/issues/36789

### Component(s)

Python

### PR DESCRIPTION
**Pull request by js8544**
### Rationale for this change

I forgot to cast durations to doubles in the current `division(duration, duration)` kernel. So they were essentially `reinterpret_cast`ed  to double. Because I only tested small positive ints but not large ints or negative ints, I missed this bug.

### What changes are included in this PR?

Add a `FloatingDivide` operator that casts ints to doubles and do floating division. Replace the `division(duration, duration)` with this op.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.

* Closes: #39156

### CODE REVIEW

**Comment by rohanjain101** [comment]
Does this fix https://github.com/apache/arrow/issues/38566 as well?

**Comment by js8544** [comment]
> Does this fix #38566 as well?

Yes! It should. Sorry for missing your issue.

File: cpp/src/arrow/compute/kernels/base_arithmetic_internal.h [hunk 1]
```diff
@@ -426,6 +426,45 @@ struct DivideChecked {
...
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_floating_value<Arg0> Call(KernelContext*, Arg0 left, Arg1 right,
+                                             Status* st) {
+    static_assert(std::is_same<T, Arg0>::value && std::is_same<T, Arg1>::value);
+    if (ARROW_PREDICT_FALSE(right == 0)) {
+      *st = Status::Invalid("divide by zero");
+      return 0;
+    }
+    return left / right;
+  }
```
Comment (kou) [inline cpp/src/arrow/compute/kernels/base_arithmetic_internal.h:463]: How about using the same order as `FloatingDivide` for easy to read?
(`enable_if_floating_value` -> `enable_if_integer_value` order.)

**Review by kou — APPROVED** [review]
+1

**Pull request merged**